source_gen/lib/src/utils.dart

```diff
@@ -94,6 +99,12 @@ Uri normalizeDartUrl(Uri url) => url.pathSegments.isNotEmpty
     ? url.replace(pathSegments: url.pathSegments.take(1))
     : url;
 
+Uri fileToAssetUrl(Uri url) {
+  if (!p.isWithin(p.current, url.path)) return url;
```

Comment:
Perhaps we could do something with the full package config to generalize this a bit more? For instance we could iterate over the package roots instead of assuming that `p.current` is within any package.